### PR TITLE
Zealous Zebras

### PR DESCRIPTION
**Pull request by DavidLopezSaez**

### CODE REVIEW

**Comment by SebastiaanZ** [comment]
Merge notes:
There's a commit that was pushed two hours after the deadline; we'll need to figure out what to do with it. Another thing to note is that there's another PR open that's related to this team (#67). It appears to be a PR that should have been made internally to the fork but somehow ended up here.

---

No flake8 linting errors.

**Comment by SebastiaanZ** [comment]
I reverted the late commit and pushed that revert to have a PR diff that shows the correct commits. That's handy for the review. The failing check related to this merge should be ignored; it's caused by other team submissions that were already merged.

File: zealous-zebras/timescape/engine/models.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+from django.db import models
```
Comment (eivl) [inline zealous-zebras/timescape/engine/models.py:1]: nice use of `__repr__`

File: zealous-zebras/timescape/engine/viewsets.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        It first looks if there is a recent search with the same query on the database.
+        If it is already in the database it returns the related results.
+        If it is not in the database it will call 'Google Custom Search JSON API'
+        (https://developers.google.com/custom-search/v1/overview) to get the results + store them + create a response
+        with new stored results.
+
+        query_params:
+            - query: required, search query
+            - page: page that should be returned. Default: 1 - Max: 5
+    '''
```
Comment (eivl) [inline zealous-zebras/timescape/engine/viewsets.py:27]: nice docstring with useful clearification.

File: zealous-zebras/timescape/engine/viewsets.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    @action(detail=False, methods=['get'])
+    def search(self, request):
+        try:
+            query = request.query_params['query']
+        except MultiValueDictKeyError:
+            return Response({'status': 'bad request'}, status=status.HTTP_400_BAD_REQUEST)
```
Comment (eivl) [inline zealous-zebras/timescape/engine/viewsets.py:33]: nice idiomatic use of try/except

File: zealous-zebras/timescape/engine/viewsets.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                        results.append(result)
+                        search.results.add(result, through_defaults={'page': page})
+
+            serializer = SearchResultSerializer(results, many=True)
+            return Response(serializer.data)
+        return Response({'status': 'no results found'})
+
+    def get_new_results(self, start_index, query):
+        with open('tokens.json', 'r') as f:
+            tokens = json.load(f)
```
Comment (eivl) [inline zealous-zebras/timescape/engine/viewsets.py:77]: nice use of a context manager.

File: zealous-zebras/timescape/engine/viewsets.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                    search_result_serializer = SearchResultSerializer(data=r)
+
+                    if search_result_serializer.is_valid():
+                        result = search_result_serializer.save()
+                        results.append(result)
+                        search.results.add(result, through_defaults={'page': page})
+
+            serializer = SearchResultSerializer(results, many=True)
+            return Response(serializer.data)
+        return Response({'status': 'no results found'})
```
Comment (eivl) [inline zealous-zebras/timescape/engine/viewsets.py:73]: It feels very "meaty" and nested. But it is very clear.

File: zealous-zebras/timescape/frontend/README.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
+This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app).
```
Comment (eivl) [inline zealous-zebras/timescape/frontend/README.md:1]: well made README

**Review by eivl — COMMENTED** [review]
The projects use of docker-compose, npm, react and pip to setup everything at runtime is really well done. It feels like a well planned stack and it makes it easy for me to setup and start. It has a special level of professionalism that you see in seasoned developers. The use of a backend and fronend docker file looks very well made. 

The Readme looks great, lots of details. 
The use of `__repr__` and f-strings makes me think that this team has control over debugging. 
There are not many docstrings, but when it's used its really well made. 

The project shows good understanding of Django, class based views, and modern python features. Some part of the viewset.py feels a bit "fat" and nested. it might have been better to refacor it down some, but the overall use of nice exception handling makes me forgive you guys :D

Code quality is far over average, good job.

**Pull request merged**